draft-ietf-httpbis-early-hints.md

```diff
@@ -90,21 +90,22 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
 The 103 (Early Hints) informational status code indicates to the client that the server is likely to
 send a final response with the header fields included in the informational response.
 
-A server MUST NOT include Content-Length, Transfer-Encoding, or any hop-by-hop header fields
-([RFC7230], Section 6.1) in a 103 (Early Hints) response.
-
 A client can speculatively evaluate the header fields included in a 103 (Early Hints) response while
 waiting for the final response. For example, a client might recognize a Link header field value
 containing the relation type "preload" and start fetching the target resource.
-
 However, these header fields only provide hints to the client; they do not replace the header
-fields on the final response. Aside from performance optimizations, such evaluation of the 103
+fields on the final response.
+
+Aside from performance optimizations, such evaluation of the 103
 (Early Hints) response's header fields MUST NOT affect how the final response is processed. A
 client MUST NOT interpret the 103 (Early Hints) response header fields as if they applied to
 the informational response itself (e.g., as metadata about the 103 (Early Hints) response).
 
-An intermediary MAY drop the informational response. It MAY send HTTP/2 ([RFC7540]) server pushes
-using the information found in the 103 (Early Hints) response.
+A server MUST NOT omit sending the header fields of a final response even if they have been sent as
+part of a 103 (Early Hints) response.
+The server SHOULD only send the header fields that are likely to help the client carry out speculative operations as part of the 103 (Early Hints) response, instead of including all the header fields that are likely to be seen in the final response.
+
+An intermediary MAY send HTTP/2 ([RFC7540]) server pushes using the information found in the 103 (Early Hints) response.
```

Comment:
I'd leave it with any intermediary. A good end-to-end example would involve a client not aware of 103, a caching proxy aware of 103, a gateway (reverse proxy) sending some links in 103 roughly based on the URL and a server knowing perfectly what to prefetch. When the gateway receives the request, it would immediately emit a 103 with 2-3 links which are generic to the site (eg: css, logo) and passes the request to the server. The server knows based on the URL that some JS is needed as well and could send an extra 103 with 2-3 more links while processing. The proxy receiving these 103 can start to prefetch these objects and populate its cache, then send some push promises to the client with these objects. That way the limitations of pushed resources (not cacheable, limited to connection etc) are significantly outcome as the client-side caching proxy can provide them very quickly again.